### ISSUE TITLE
minetest.serialize produces bad output for some tables with circular references

### ISSUE THREAD

**Issue by FaceDeer** [minetest/minetest#8719]
##### Minetest version

```
Minetest 5.0.1
```

##### OS / Hardware

Operating system: Windows 10 Home
CPU: Intel i7-3770K @ 3.5GHz, 8.00 GB ram

##### Summary

In the course of attempting to persist some tables between runs I've discovered that minetest.serialize() produces broken output for some (but not all) tables with circular references. The resulting output crashes with a nil dereference error when attempting to loadstring() it back into a table again; I'll detail a reliable example in the "steps to reproduce" section below.

This doesn't occur in all cases, I've successfully serialized and deserialized tables with circular references before. It appears that a particular pattern of nested circular references are required to trigger this bug.

##### Steps to reproduce

The following code can be used as a standalone init.lua mod file to illustrate the problem:

```
-- Initial non-circular-reference-containing data structure:
local data = {
	player_accounts = {
		player1 = {
			log = {
				{}
			},
		},
		player2 = {
			log = {},
		},
	},
}

-- Adding the circular references. The two player accounts have a log entry that
-- they share a reference to, and the log entry references both player accounts.
data.player_accounts.player1.log[1].seller = data.player_accounts.player1
data.player_accounts.player1.log[1].purchaser = data.player_accounts.player2
data.player_accounts.player2.log[1] = data.player_accounts.player1.log[1]

local serialized = minetest.serialize(data)
minetest.debug("Source data:")
minetest.debug(dump(data))
minetest.debug("Serialized data:")
minetest.debug(serialized)
-- minetest.serialize outputs Lua code that, when executed, builds and returns
-- the original data structure. minetest.deserialize() simply returns nil on
-- failure, so I manually loadstring() and execute the result here to make
-- the error explicit:
local loadedstring = loadstring(serialized)
local data2 = loadedstring() -- This step crashes
minetest.debug("Loadstring data:")
minetest.debug(dump(data2))
```

This is the serialized version of the ``data`` structure that minetest.serialize produces, with added comments showing where it's gone wrong:

```
local _ = {}
_[1] = {}
_[2] = {["log"] = _[1]}
_[3] = {}
_[4] = {["log"] = _[3]}
-- There should be "_[5] = {}" at this line, but there is not.
_[3][1] = _[5] -- index 5 is nil at this point so this line does nothing
_[5]["purchaser"] = _[2] -- this line crashes with an attempt to index nil.
_[5]["seller"] = _[4]
_[1][1] = _[5]
return {["player_accounts"] = {["player2"] = _[2], ["player1"] = _[4]}}
```

**Comment by DS-Minetest** [comment]
<details><summary>I've done some modifications to the code to remove some unneeded stuff and to give the tables names.</summary>

```lua
-- Initial non-circular-reference-containing data structure:
local data = {
	name = "data",
	A = {
		name = "A",
		At = {
			name = "At",
			{name = "At1"}
		},
	},
	B = {
		name = "B",
		Bt = {name = "Bt"},
	},
}

-- Adding the circular references. The two player accounts have a log entry that
-- they share a reference to, and the log entry references both player accounts.
data.A.At[1].Ai = data.A
data.A.At[1].Aii = data.B
data.B.Bt[1] = data.A.At[1]
minetest.log("Source data:\n"..dump(data))

local serialized = minetest.serialize(data)
minetest.log("Serialized data:\n"..serialized)

local data_out, err = minetest.deserialize(serialized)
if not data_out then
	minetest.log("error", "deserialize error: "..err)
	return
end
minetest.log("Loadstring data:\n"..dump(data2))
```
</details><details><summary>This is what it says for the source data</summary>

```
2019-07-28 13:29:48: [Main]: Source data:
{
	B = {
		name = "B",
		Bt = {
			{
				Aii = <circular reference>,
				name = "At1",
				Ai = {
					name = "A",
					At = {
						<circular reference>,
						name = "At"
					}
				}
			},
			name = "Bt"
		}
	},
	name = "data",
	A = {
		name = "A",
		At = {
			{
				Aii = {
					name = "B",
					Bt = {
						<circular reference>,
						name = "Bt"
					}
				},
				name = "At1",
				Ai = <circular reference>
			},
			name = "At"
		}
	}
}
```
</details><details><summary>and for the serialized data.</summary>

```lua
2019-07-28 13:29:48: [Main]: Serialized data:
local _ = {}
_[1] = {["name"] = "Bt"}
_[2] = {["name"] = "B", ["Bt"] = _[1]}
_[3] = {["name"] = "At"}
_[4] = {["name"] = "A", ["At"] = _[3]}
_[5]["Aii"] = _[2]
_[5]["Ai"] = _[4]
_[3][1] = _[5]
_[1][1] = _[5]
return {["B"] = _[2], ["name"] = "data", ["A"] = _[4]}
```
</details><details><summary>And here is the error.</summary>

`2019-07-28 13:29:48: ERROR[Main]: deserialize error: (load):6: attempt to index a nil value`
</details>

**Comment by FaceDeer** [comment]
In case anyone else encounters this problem I figured I should mention my current workaround; I've included [this Lua table persistence code](https://the-color-black.net/blog/article/LuaTablePersistence) in my mod and am using it instead of minetest.serialize() for the time being. It's under the MIT license and doesn't appear to share the bug reported here. There are [tons and tons of other similar bits of code out there](http://lua-users.org/wiki/TableSerialization) in case this one turns out to be problematic as well.
